### PR TITLE
Unify Beluga documentation

### PR DESCRIPTION
**Pull request by hidmic**
### Proposed changes

This patch unifies all repository documentation in a common Sphinx site: high-level documentation lives in [MyST](https://mystmd.org/) flavor Markdown documents, Doxygen API documentation is pulled in using the `autodox` extension in https://github.com/Ekumen-OS/sphinx-babel. 

In the process, this PR partially addresses #301, #302, #303, #304, and #308. 

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [ ] 🚀 Feature (change which adds functionality)
- [x] 📚 Documentation (change which fixes or extends documentation)

💥 **Breaking change!** This patch changes the way we used to build documentation for the Beluga package.

### Checklist

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### Additional comments

This patch gets us 80% of the way towards release ready documentation. High-level documentation is not as polished as I would like to (specially the Key Concepts page), but this patch is big enough already and I'm too drained to produce quality technical writing. The easiest way to review this is building documentation locally and exploring it. You can check the README file under `docs` for instructions, or simply:

```
git clone https://github.com/Ekumen-OS/beluga.git
cd beluga/docs
python3 -m venv venv
source venv/bin/activate
pip install -r requirements.txt
make html 
```

Note this documentation needs Python 3.9 or later to build. I was not able to get all dependencies to play ball in earlier Python versions.

### CODE REVIEW

File: DEVELOPING.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Developing Beluga
```
Comment (hidmic) [inline DEVELOPING.md:1]: Note that some repository-level documentation has been kept around. What looks good in Github doesn't look good in Sphinx and viceversa. The idea here is to keep these READMEs around as an addendum to code for direct contributors.

File: docs/packages/beluga [hunk 2]
```diff
@@ -0,0 +1 @@
+../../beluga
```
Comment (hidmic) [inline docs/packages/beluga:1]: Note the symlinks. To keep package documentation as close as possible to package code, I had to pull in packages under the Sphinx source directory. I don't love it but some of the Sphinx extensions we use leave little margin of action.

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+```
+
+```{toctree}
+:hidden:
+:maxdepth: 1
+:caption: Packages
+
+beluga <packages/beluga/docs/index>
+beluga_ros <packages/beluga_ros/docs/index>
+beluga_amcl <packages/beluga_amcl/docs/index>
```
Comment (hidmic) [inline docs/index.md:82]: Note that not all packages have been added to the documentation. `beluga_system_tests` and `beluga_example` seemed accessory, `beluga_benchmark` should be here but it's a bit behind our latest benchmarks. Plan is to pull it in as soon as we can migrate LAMBKIN stuff to it (WIP).

File: docs/tutorials/particle-filtering.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Primer on Particle Filtering with Beluga
+
+Coming soon!
```
Comment (hidmic) [inline docs/tutorials/particle-filtering.md:3]: @Alondruck I left this placeholder for the one-dimensional particle filter tutorial (#340).

File: docs/tutorials/nav2-integration.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Using Beluga with Nav2
```
Comment (hidmic) [inline docs/tutorials/nav2-integration.md:1]: @Alondruck I left this placeholder for the Nav2 integration tutorial (#307).

File: docs/guides/extending-beluga.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Extending Beluga
+
+Coming soon!
```
Comment (hidmic) [inline docs/guides/extending-beluga.md:3]: @nahueespinosa I left this placeholder for the Beluga extension guide (#306).

File: docs/requirements.txt
Comment (nahueespinosa) [inline docs/requirements.txt:1]: It seems that we also need:
https://pypi.org/project/sphinxcontrib-pseudocode/

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+> [!TIP]
+> Use a [virtual environment](https://docs.python.org/3/library/venv.html) to avoid polluting your system.
+
+## Build HTML documentation
+
+Simply make it!
+
+```sh
+make
```
Comment (nahueespinosa) [inline docs/README.md:25]: I tried this, but there is something wrong with the `sphinx_babel` extension:
```
Running Sphinx v7.3.7
checking bibtex cache... out of date
parsing bibtex file /home/nahue/ekumen/beluga/docs/references.bib... parsed 6 entries
WARNING: multiple files found for the document "venv/lib/python3.10/site-packages/sphinx_mdinclude/tests/test": ['venv/lib/python3.10/site-packages/sphinx_mdinclude/tests/test.md', 'venv/lib/python3.10/site-packages/sphinx_mdinclude/tests/test.rst']
Use '/home/nahue/ekumen/beluga/docs/venv/lib/python3.10/site-packages/sphinx_mdinclude/tests/test.md' for the build.
WARNING: multiple files found for the document "venv/lib64/python3.10/site-packages/sphinx_mdinclude/tests/test": ['venv/lib64/python3.10/site-packages/sphinx_mdinclude/tests/test.md', 'venv/lib64/python3.10/site-packages/sphinx_mdinclude/tests/test.rst']
Use '/home/nahue/ekumen/beluga/docs/venv/lib64/python3.10/site-packages/sphinx_mdinclude/tests/test.md' for the build.
[autodox] processing beluga

Extension error (sphinx_babel.autodox):
Handler <function generate_doxygen_documentation at 0x77dea0c3c040> for event 'builder-inited' threw an exception (exception: 'HTML_OUTPUT')
make: *** [Makefile:26: html] Error 2
```

**Review by nahueespinosa — COMMENTED** [review]
@hidmic This is absolutely amazing! I have a couple of quick questions about the setup before I start the review.

File: docs/requirements.txt
Comment (hidmic) [reply docs/requirements.txt:1]: Ah, no, that's a leftover in the `conf.py` file. `sphinxcontrib-pseudocode` is somewhat broken for latest Sphinx versions.

File: docs/README.md (hunk 7 above)
Comment (hidmic) [reply docs/README.md:25]: Hmm, this looks like Doxygen specific. Which version are you running?

Comment (nahueespinosa) [reply docs/README.md:25]: Ah, apparently none on my host machine :sweat_smile:.

This is what I see now:

```
❯ make html
Running Sphinx v7.3.7
checking bibtex cache... out of date
parsing bibtex file /home/nahue/ekumen/beluga/docs/references.bib... parsed 6 entries
[autodox] processing beluga
warning: ignoring unsupported tag 'ST' at line 65, file /home/nahue/ekumen/beluga/docs/packages/beluga/docs/_doxygen/generated/reference/beluga.Doxyfile
warning: tag INPUT: input source './_doxygen/main.md' does not exist
Doxygen version used: 1.9.1
Searching for include files...
Searching for example files...
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/actions
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/algorithm
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/algorithm/raycasting
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/containers
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/include
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/include/beluga
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/include/beluga/test
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/include/beluga/test/utils
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/motion
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/policies
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/random
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/sensor
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/sensor/data
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/test_data
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/testing
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/type_traits
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/utility
Searching for files in directory /home/nahue/ekumen/beluga/beluga/test/beluga/views
Searching for images...
Searching for dot files...
Searching for msc files...
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/actions
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/algorithm
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/algorithm/raycasting
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/containers
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/detail
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/motion
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/policies
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/random
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/sensor
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/sensor/data
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/testing
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/type_traits
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/types
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/utility
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/beluga/views
Searching for files in directory /home/nahue/ekumen/beluga/beluga/include/ciabatta
error: source './_doxygen/main.md' is not a readable file or directory... skipping.
 (warning treated as error, aborting now)
```

Comment (hidmic) [reply docs/README.md:25]: Ah, an older `.gitignore` rule. Fixed in https://github.com/Ekumen-OS/beluga/pull/346/commits/9e63ebdd415185a64b2efc39be481a789de20367.

Comment (hidmic) [reply docs/README.md:25]: It does not. Doxygen version?

Comment (nahueespinosa) [reply docs/README.md:25]:
```bash
doxygen --version
1.9.1
```

Comment (hidmic) [reply docs/README.md:25]: Indeed. Tagfiles in Doxygen 1.9.1 include (or changed?) page anchors, producing the multiple definitions you see above.  Fixed in https://github.com/Ekumen-OS/beluga/pull/346/commits/0a28f43254cf5a5c26695749661533258d26c907.

Comment (nahueespinosa) [reply docs/README.md:25]: Success! Thank you.

**Comment by hidmic** [comment]
Looks like Rolling moved on to Noble already and we are getting build failures :eyes:

**Review by nahueespinosa — COMMENTED** [review]
@hidmic I love this! First pass, I'll continue the review later.

File: README.md [hunk 8]
```diff
@@ -44,5 +44,5 @@ This repository contains the following packages:
 
 - Get hands-on experience with the [getting started](GETTING_STARTED.md) tutorial.
 - Check the [installation guide](INSTALLING.md) to test this on a robot.
```
Comment (nahueespinosa) [inline README.md:46]: Maybe remove these two lines as the links don't work anymore.
```suggestion
```

File: GETTING_STARTED.md [hunk 9]
```diff
@@ -1,84 +0,0 @@
...
-    ```
-
-    For ROS 1 distributions, in two separate terminals run:
-    ```bash
-    cd /ws
-    source devel*/setup.bash
-    roslaunch beluga_example simulation.launch
-    ```
-    ```bash
-    rosrun teleop_twist_keyboard teleop_twist_keyboard
```
Comment (MARIANOCEREDA) [inline GETTING_STARTED.md:71]: If the steps in this file are followed, when tried running with **ROS Noetic**, there's only a **.py** file to run teleop_twist_keyboard.
Then, it only works if the command is : `rosrun teleop_twist_keyboard teleop_twist_keyboard.py`.

Comment (hidmic) [reply GETTING_STARTED.md:71]: Thanks! Fixed in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11

File: README.md (hunk 8 above)
Comment (hidmic) [reply README.md]: Removed in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11.

File: docs/concepts/key-concepts.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+and an update step. During the prediction step a new set of samples is generated using a proposal distribution that includes the information of the previous belief and the effect of the current control inputs:
+
+$$
+    q(x_t|x_{t-1}, u_t) = p(x_t|x_{t-1}, u_t) bel(x_{t-1})
+$$
+
+After this step the resulting particle set distribution does not match the updated belief (target distribution), since the proposal distribution does not match the target distribution in general. The resampling step is used to correct this mismatch. In this step each particle is assigned an importance weight which is calculated from the sensor readings and the particle state updated by the prediction step.
```
[the remark this answers is not in the document]
Comment (hidmic) [reply docs/concepts/key-concepts.md]: Changed in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11

File: docs/concepts/key-concepts.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+$$
+    w^{[i]}_t = p(z_t|x^{[i]}_t
```
[the remark this answers is not in the document]
Comment (hidmic) [reply docs/concepts/key-concepts.md]: Fixed in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11.

File: docs/concepts/key-concepts.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+$$
+    bel(x_t) = p(x_t|z_{1:t}, u_{1:t})
+$$
+
+where $x_t$ is the state of the system at time $t$, and $z_{1:t}$ and $u_{1:t}$ are the sequence of sensor measurements and the sequence of control inputs up to time $t$ respectively.
+
+It can be shown that the posterior belief can be recursively computed using the prior belief, the current sensor readings, and the current control inputs using the following update rule:
+
+$$
+    bel(x_t) = η p(z_t|x_t) \int p(x_t|x_{t-1}, u_t) bel(x_{t-1}) dx_{t-1}
```
[the remark this answers is not in the document]
Comment (hidmic) [reply docs/concepts/key-concepts.md]: Done in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11

File: docs/getting-started/installation.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+:::{tab-item} Ubuntu
+```bash
+source /opt/ros/${ROS_DISTRO}/setup.bash
+```
+:::
+
+::::
+
+If in doubt as to which ROS distribution to use, check the list of [supported distributions](../packages/beluga_ros/docs/index).
```
[the remark this answers is not in the document]
Comment (hidmic) [reply docs/getting-started/installation.md]: Ahh, good catch. Fixed in https://github.com/Ekumen-OS/beluga/pull/346/commits/3e9961897fcd9f2f23db5253b2d018be630b0f11.

**Review by nahueespinosa — APPROVED** [review]
LGTM!

File: .github/workflows/ci_pipeline.yml [hunk 14]
```diff
@@ -107,14 +103,38 @@ jobs:
...
+
+      - name: Setup python
+        uses: actions/setup-python@v4
+        with:
+          cache: pip
+
+      - name: Install tooling dependencies
+        run: >-
+          sudo apt-get update &&
+          sudo apt-get install -y doxygen git make &&
```
Comment (nahueespinosa) [inline .github/workflows/ci_pipeline.yml:121]: We might be missing a package here according to doc dependencies: https://github.com/Ekumen-OS/beluga/blob/main/beluga/package.xml#L31C15-L31C33

Maybe `texlive-latex-base` and `graphviz`?

```
error opening bib2759.bbl: No such file or directory
Preprobibtex command failed: No such file or directory
error: Problems running bibtex. Verify that the command 'perl --version' works from the command line. Exit code: 2
```

Comment (hidmic) [reply .github/workflows/ci_pipeline.yml:121]: Indeed. I may add a new Docker image in a separate PR. `texlive` is large.

Comment (hidmic) [reply .github/workflows/ci_pipeline.yml]: Alright, `texlive-latex-base` seems small enough for now. I'll defer more infra work to https://github.com/Ekumen-OS/beluga/issues/354.

**Comment by hidmic** [comment]
Rolling CI ifailure will be fixed by https://github.com/open-navigation/navigation2/pull/4267.

**Pull request merged**